pkg/postage/service.go

```diff
@@ -165,24 +143,35 @@ func (ps *service) IssuerUsable(st *StampIssuer) bool {
 func (ps *service) GetStampIssuer(batchID []byte) (*StampIssuer, func() error, error) {
 	ps.lock.Lock()
 	defer ps.lock.Unlock()
-	for _, st := range ps.issuers {
-		if bytes.Equal(batchID, st.data.BatchID) {
-			if !ps.IssuerUsable(st) {
-				return nil, nil, ErrNotUsable
-			}
-			return st, func() error {
-				return ps.save(st)
-			}, nil
-		}
+
+	if _, ok := ps.issuerLocks[string(batchID)]; !ok {
+		ps.issuerLocks[string(batchID)] = &sync.Mutex{}
+	}
+	ps.issuerLocks[string(batchID)].Lock()
+
+	item := NewStampIssuerItem(batchID)
+	err := ps.store.Get(item)
+	if errors.Is(err, storage.ErrNotFound) {
+		return nil, nil, ErrNotFound
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if !ps.IssuerUsable(item.Issuer) {
+		return nil, nil, ErrNotUsable
 	}
-	return nil, nil, ErrNotFound
+	return item.Issuer, func() error {
+		defer ps.issuerLocks[string(batchID)].Unlock()
```

Comment:
If you use the empty struct, I think you will need the lock here as save doesnt acquire the lock.